### PR TITLE
Add SSH key support (cont. for #5154)

### PR DESCRIPTION
**Pull request by bcho**
###### Before submitting the PR, please make sure you do the following

- [ ] If you're attempting to fix a translation issue, please submit your changes to our [POEditor project](https://poeditor.com/join/project/ynjQmwSsGh) instead of opening a PR.

### Tests
-  [ ] Run the frontend and integration tests with  `yarn lint && yarn flow && yarn test`)
-  [x] If there are changes to the backend codebase, run the backend tests with `lein test && lein lint && ./bin/reflection-linter`

-  [x] Sign the [Contributor License Agreement](https://docs.google.com/a/metabase.com/forms/d/1oV38o7b9ONFSwuzwmERRMi9SYrhYeOrkbmNaq9pOJ_E/viewform)
(unless it's a tiny documentation change).

------------------------

This pull request is the test follow up for #5154:

- added sshd as mock sssh server
- added ssh connection related test cases (password, key-based challenge)

### CODE REVIEW

**Comment by bcho** [comment]
CLA signed! Please have a look @camsaul

File: project.clj
Comment (bcho) [inline project.clj]: This is a test dependency, I didn't find a proper way to put it cc @camsaul

File: test_resources/ssh_test
Comment (bcho) [inline test_resources/ssh_test]: these keys were generated by `ssh-keygen -t rsa  -m PEM`

File: project.clj
Comment (camsaul) [reply project.clj]: The `:dev` profile is probably the best place to put it

Comment (bcho) [reply project.clj]: Updated, PTAL @camsaul

**Review by flyingmachine — CHANGES_REQUESTED** [review]
This is great, thank you! The comments are mostly minor requests for changes :) There is one larger question for @mazameli about how to handle the SSH settings in the UI.

**Comment by bcho** [comment]
That's wired, I don't know why tests keep failing (no `more-of`). But last time I tested, it worked on circleci... @flyingmachine any idea? (The last time I wrote clojure was 2 years ago, I don't know is anything changed in that time, thanks in advance!)

**Comment by bcho** [comment]
> There is one larger question for @mazameli about how to handle the SSH settings in the UI.

I don't have any suggestions here since I don't use metabase very often. But I think [sequel pro](https://www.sequelpro.com/) provides some great examples on how to handle SSH login (with or without ssh key pass phrase).

File: src/metabase/util/ssh.clj
Comment (flyingmachine) [reply src/metabase/util/ssh.clj]: @mazameli would love your thoughts here :)

**Comment by flyingmachine** [comment]
@bcho I've updated the tests so that they pass now :) The issue was with how our testing approach has evolved to move from `expectations` to `clojure.test`. Thank you for your contribution!

Now we're just waiting on feedback on how to handle this in the UI.

**Review by flyingmachine — APPROVED** [review]

**Comment by mazameli** [comment]
@flyingmachine If I'm understanding correctly the the password vs. the SSH key are mutually exclusive, then I think the cleanest solution would be to ask which auth method the user wants to use, and then making the password or SSH key field visible and required based on the selection. (I've indented these SSH settings in the mocks to clarify that they're sub-settings of the SSH-tunnel toggle, but that's a nicety, not a requirement.)

**Default:**

image

**If SSH key is selected:**

image

**Review by mazameli — CHANGES_REQUESTED** [review]
I added some suggestions last week (above) for how to make the choice between using a key vs. a password clearer. @flyingmachine, can you let me know if the suggestions make sense to you? If so, then we'll need to ask @bcho to implement that change before we merge this.

**Comment by mazameli** [comment]
@bcho, I confirmed with @flyingmachine that my [change requests](https://github.com/metabase/metabase/pull/10758#issuecomment-539252180) make sense to him. If you're able to make that change, we'd happily accept this PR. Thanks!

**Comment by bcho** [comment]
OK I will have a look this weekend

**Comment by bcho** [comment]
@mazameli @flyingmachine 

I am not quite following what exactly the changes I have to do for the UI. Do I need to add a new input field (which render the radio widget like the mockup) in [here](https://github.com/metabase/metabase/blob/18da2fe56cfb6a2bc510ee6755f558a72c964e52/frontend/src/metabase/components/DatabaseDetailsForm.jsx#L153-L180)?

If it's true, I think it's better to split the tunnel rendering part out and render them like what the mockup do (because I don't find any examples for rendering fields like the suggested UI). Is that OK to do so?

**Comment by flyingmachine** [comment]
@bcho thanks for having a look! As I understand it, the work would be to add This field:

```
How do you want to authenticate?
[ x ] Password
[    ] SSH key
```

Password and SSH key would be radio buttons. If you select password, then the UI would show a text input for the password: 

```
How do you want to authenticate?
[ x ] Password
[    ] SSH key

Password:
[ password input ]
```

If you select SSH key, then the UI would show a text field for the path to the key:

```
How do you want to authenticate?
[ x ] Password
[    ] SSH key

Path on the Metabase server to a SSH private key file:
[ text input for path ]
```

Does this help?

**Comment by robdaemon** [comment]
I'm closing this in favor of #12548 - thank you for the patch! It got this feature going.

**Pull request closed without merging**
